enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,412 @@
+related resources while that Machine's reconciliation is temporarily paused.
+This pause in reconciliation will allow these custom components to take action
+after a Machine has been marked for deletion, but prior to the Machine being
+drained and/or associated instance terminated.
+
+In particular, this will allow the Etcd Operator to prevent a Control Plane
+Machine from being drained/removed until the Etcd Operator has had a chance
+to synchronise the Etcd data to the replacement Machine.
+
+This could also be used by other components such as Storage to allow them to
+ensure safe detachment of volumes before the Machine is deleted.
+
+### Goals
+
+- Define an initial set of hook points for the deletion phase.
+- Define an initial set and form of related annotations.
+- Define basic expectations for a controller or process that responds to a
+lifecycle hook.
+
+### Non-Goals
+
+- Create an exhaustive list of hooks; we can add more over time.
+- Create new machine phases.
+- Create a mechanism to signal what lifecycle point a machine is at currently.
+- Dictate implementation of controllers that respond to the hooks.
+- Implement ordering in the machine controller.
+- Require anyone to use these hooks for normal machine operations, these are
+strictly optional and for custom integrations only.
+
+## Proposal
+
+- Utilize annotations to implement lifecycle hooks.
+- Each lifecycle point can have 0 or more hooks.
+- Hooks do not enforce ordering.
+- Hooks found during machine reconciliation effectively pause reconciliation
+until all hooks for that lifecycle point are removed from a machine's annotations.
+
+### User Stories
+
+#### Story 1
+(pre-terminate) As an operator, I would like to have the ability to perform
+different actions between the time a Machine is marked deleted in the API and
+the time the Machine is deleted from the infrastructure provider.
+
+For example, when replacing a Control Plane Machine, ensure a new Control
+Plane Machine has been successfully created and joined to the cluster before
+removing the instance of the deleted Machine. This might be useful in case
+there are disruptions during replacement and we need the disk of the existing
+instance to perform some disaster recovery operation.  This will also prevent
+prolonged periods of having one fewer Control Plane host in the event the
+replacement instance does not come up in a timely manner.
+
+#### Story 2
+(pre-terminate) As an operator, I would to have the ability to perform some
+action between the time that a Machine has been drained, but before it is
+removed from the infrastructure provider.
+
+For example, when storage is attached to the Machine due to some pod placement,
+an operator may want to check that the storage is detached before allowing the
+removal of the machine from the infrastructure provider.
+
+#### Story 3
+(pre-drain) As an operator, I want the ability to utilize my own draining
```

Comment:
Do we know of a specific real need for this? I'd argue this can possibly be solved by exposing drain parameters?

Reply:
This possibly could be exposed by exposing drain parameters, though this is something we don't have today. I think this was also an example copied from the upstream enhancement so I'm not fully up to date on what the exact use case was for this.

Perhaps the idea is that this is generic enough that someone could utilize their own drain logic without CAPI/MAPI having to maintain the complexity of an exposed drain controller?